### PR TITLE
FFI: Support the use of via parameters when looking up and joining rooms.

### PR DESCRIPTION
**Pull request by pixlwave**
This PR makes the following changes:
- When resolving a room alias, also return the server names so they can be used later if needed.
- Expose `join_room_by_alias_or_id` which allows passing the server names from the via parameters in a permalink.

It doesn't help with getting the room preview which also needs to support the via parameters, but I haven't figured out how that one works yet.

### CODE REVIEW

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 1]
```diff
@@ -725,6 +725,21 @@ impl Client {
         Ok(Arc::new(Room::new(room)))
     }
```
Comment (pixlwave) [inline bindings/matrix-sdk-ffi/src/client.rs:726]: I'm not sure if there's any benefit of leaving this `join_room_by_id` method here or not. It seems redundant to me having both, but then the SDK supports both so 🤷‍♂️

Comment (poljar) [reply bindings/matrix-sdk-ffi/src/client.rs:726]: Well the spec supports both so we have it in the SDK.

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 2]
```diff
@@ -725,6 +725,21 @@ impl Client {
...
 
+    pub async fn join_room_by_id_or_alias(
```
Comment (poljar) [inline bindings/matrix-sdk-ffi/src/client.rs:728]: Don't you want to add some docs, now that uniffi forwards them through the FFI?

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 3]
```diff
@@ -789,6 +807,21 @@ impl From<NotificationProcessSetup> for MatrixNotificationProcessSetup {
     }
 }
 
+#[derive(uniffi::Record)]
+pub struct ResolvedRoomAlias {
```
Comment (poljar) [inline bindings/matrix-sdk-ffi/src/client.rs:811]: Seems sensible, but some docs would be nice as well.

**Review by poljar — CHANGES_REQUESTED** [review]
Makes sense, but we might want to start requiring docs in the bindings.

**Comment by pixlwave** [comment]
Oh yes, I forgot we had docs now, thanks for the reminder! Added.

File: bindings/matrix-sdk-ffi/src/client.rs (hunk 1 above)
Comment (pixlwave) [reply bindings/matrix-sdk-ffi/src/client.rs:731]: Sorry yes, I'm just not sure if there's an advantage of using this one through the FFI and was thinking "out loud". Will leave them both - the docs will help clarify.

Comment (poljar) [reply bindings/matrix-sdk-ffi/src/client.rs:731]: Yeah, I mean somebody else might be using the bindings and might find it useful. So I think it makes sense to just leave it there.

**Review by poljar — APPROVED** [review]

**Pull request merged**